### ISSUE TITLE
Doctrine Source Doesn't Include Associations

### ISSUE THREAD

**Issue by mesour** [mesour/sources#10]
@oddz commented on [Sun Mar 13 2016](https://github.com/mesour/DataGrid/issues/54)

The doctrine source doesn't recure through associations when converting entities to flat arrays. This results in any of those columns not being compatible with ordering and filtering. The fix is to have the conversion process in the source not only check for fields but associations as well and include them recursively.

---

@oddz commented on [Sun Mar 13 2016](https://github.com/mesour/DataGrid/issues/54#issuecomment-195842953)

This method is the culprit.

Mesour\Source\DoctrineSource::getEntityArrayAsArrays

In that method associations also need to be mapped and flattened recursively.

Also need to add some type of naming convention for accessing that data I believe.

---

@mesour commented on [Sun Mar 13 2016](https://github.com/mesour/DataGrid/issues/54#issuecomment-195961794)

I think I will do on the entity an obligation to have a method toArray and method Mesour\Source\DoctrineSource::getEntityArrayAsArrays can be removed

---

@oddz commented on [Fri Mar 18 2016](https://github.com/mesour/DataGrid/issues/54#issuecomment-198508799)

That is an option but it would be nice to have the data source take care of that logic. Also that doesn't fix the filter and sorting incompatibility. Both of those classes validate the data based on a column with a matching key existing at the root of the array. Instead the discovery mechanism needs to account for nested arrays. Thus column names need to allow for something like author.name – where author is a nested array with a name key.

---

@mesour commented on [Fri Mar 18 2016](https://github.com/mesour/DataGrid/issues/54#issuecomment-198563879)

Oh you mean this :) Can use second parameter of `DoctrineGridSource::__construct($queryBuilder, array $columnMapping)`.

Example column mapping:

``` php
$columnMapping = [
    // key in data array => real name
    'id' => 'user.id',
    'group_name' => 'group.name',
];
$source = new DoctrineGridSource($queryBuilder, $columnMapping);
```

---

@mesour commented on [Fri Mar 18 2016](https://github.com/mesour/DataGrid/issues/54#issuecomment-198564405)

For problem with DoctrineGridSource::getEntityArrayAsArrays can use this fast fix for now:

``` php
class GridSource extends DoctrineGridSource
{
    protected function getEntityArrayAsArrays($results)
    {
        $output = [];
        foreach ($results as $result) {
            if (method_exists($result, 'toArray')) {
                $output[] = $result->toArray();
            } else {
                throw new InvalidArgumentException(
                    sprintf('Entity %s must have method toArray().', get_class($result))
                );
            }
        }
        return $output;
    }
}
```

**Comment by mesour** [comment]
@oddz: fixed in version v3.0.1-beta. Now DoctrineSource require method `toArray` on entity

**Comment by oddz** [comment]
I will let you know how it works out.

Thanks
